app/src/main/java/org/jellyfin/mobile/media/MediaService.kt

```diff
@@ -0,0 +1,568 @@
+                setIconUri(mediaImageUri)
+            }.build()
+
+        return MediaItem(description, flag)
+    }
+
+    private inner class MediaQueueNavigator(mediaSession: MediaSessionCompat) :
+        TimelineQueueNavigator(mediaSession) {
+        override fun getMediaDescription(
+            player: Player,
+            windowIndex: Int
+        ): MediaDescriptionCompat {
+            return currentPlaylistItems[windowIndex]
+        }
+    }
+
+    private inner class MediaPlaybackPreparer(
+        private val exoPlayer: ExoPlayer,
+        private val appPreferences: AppPreferences
+    ) : MediaSessionConnector.PlaybackPreparer {
+
+        override fun getSupportedPrepareActions(): Long =
+            PlaybackStateCompat.ACTION_PREPARE_FROM_MEDIA_ID or
+                PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID
+
+        override fun onPrepare(playWhenReady: Boolean) {}
+
+        override fun onPrepareFromMediaId(mediaId: String, playWhenReady: Boolean, extras: Bundle?) {
+            val shouldShuffle = mediaId.endsWith(MediaItemType.Shuffle.toString())
+
+            if (shouldShuffle) {
+                currentPlaylistItems.shuffle()
+            }
+
+            val mediaItems = currentPlaylistItems.map { item ->
+                createMediaItem(item.mediaId!!)
+            }
+
+            var mediaItemIndex = 0
+
+            if (!shouldShuffle) {
+                mediaItemIndex = currentPlaylistItems.indexOfFirst { item ->
+                    item.mediaId == mediaId
+                }
+            }
+
+            exoPlayer.setMediaItems(mediaItems)
+            exoPlayer.prepare()
+            if (mediaItemIndex in 0 until currentPlaylistItems.size)
+                exoPlayer.seekTo(mediaItemIndex, 0)
+        }
+
+        override fun onPrepareFromSearch(query: String, playWhenReady: Boolean, extras: Bundle?) {}
+
+        override fun onPrepareFromUri(uri: Uri, playWhenReady: Boolean, extras: Bundle?) = Unit
+
+        override fun onCommand(
+            player: Player,
+            controlDispatcher: ControlDispatcher,
+            command: String,
+            extras: Bundle?,
+            cb: ResultReceiver?
+        ): Boolean = false
+
+        fun createMediaItem(mediaId: String): ExoPlayerMediaItem {
+            val url = "${appPreferences.instanceUrl}/Audio/${mediaId}/universal?" +
+                "UserId=${appPreferences.instanceUserId}&" +
+                "DeviceId=${DEVICE_ID_BASE64}&" +
+                "MaxStreamingBitrate=140000000&" +
+                "Container=opus,mp3|mp3,aac,m4a,m4b|aac,flac,webma,webm,wav,ogg&" +
+                "TranscodingContainer=ts&" +
+                "TranscodingProtocol=hls&" +
+                "AudioCodec=aac&" +
+                "api_key=${appPreferences.instanceAccessToken}&" +
+                "PlaySessionId=${UUID.randomUUID()}&" +
+                "StartTimeTicks=0&" +
+                "EnableRedirection=true&" +
+                "EnableRemoteMedia=true"
+
```

Comment:
Looking at https://github.com/jellyfin/jellyfin-apiclient-java/blob/master/jellyfin-api/src/main/kotlin-generated/org/jellyfin/apiclient/api/operations/UniversalAudioApi.kt#L43

- The deviceId should probably be plaintext, not base64
- PlaySessionId doesn't exist
- StartTimeTicks _should_ default to 0 when missing
- EnableRedirection is true by default